compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -1348,70 +1343,46 @@ class PatternMatcher extends MiniPhaseTransform with DenotTransformer {
         else resultType
 
-      def resultType: Type
-
-      /** Create the TreeMaker that embodies this extractor call
-        *
-        * `binder` has been casted to `paramType` if necessary
-        * `binderKnownNonNull` indicates whether the cast implies `binder` cannot be null
-        * when `binderKnownNonNull` is `true`, `ProductExtractorTreeMaker` does not do a (redundant) null check on binder
-        */
-      def treeMaker(binder: Symbol, binderKnownNonNull: Boolean, pos: Position, binderTypeTested: Type): TreeMaker
-
       // `subPatBinders` are the variables bound by this pattern in the following patterns
       // subPatBinders are replaced by references to the relevant part of the extractor's result (tuple component, seq element, the result as-is)
       // must set infos to `subPatTypes`, which are provided by extractor's result,
       // as b.info may be based on a Typed type ascription, which has not been taken into account yet by the translation
       // (it will later result in a type test when `tp` is not a subtype of `b.info`)
       // TODO: can we simplify this, together with the Bound case?
       def subPatBinders = subBoundTrees map (_.binder)
-      lazy val subBoundTrees = (args, subPatTypes).zipped map newBoundTree
+      lazy val subBoundTrees = (args, typedPatterns.map(_.tpe)).zipped map newBoundTree
 
       // never store these in local variables (for PreserveSubPatBinders)
       lazy val ignoredSubPatBinders: Set[Symbol] = subPatBinders zip args collect { case (b, PatternBoundToUnderscore()) => b } toSet
 
-      // do repeated-parameter expansion to match up with the expected number of arguments (in casu, subpatterns)
-      private def nonStarSubPatTypes = aligner.typedNonStarPatterns map (_.tpe)
-
-      def subPatTypes: List[Type] = typedPatterns map (_.tpe)
-
       // there are `prodArity` non-seq elements in the tuple.
-      protected def firstIndexingBinder = prodArity
-      protected def expectedLength      = elementArity
-      protected def lastIndexingBinder  = totalArity - starArity - 1
+      private def firstIndexingBinder = prodArity
+      private def expectedLength      = elementArity
+      private def lastIndexingBinder  = totalArity - starArity - 1
 
       private def productElemsToN(binder: Symbol, n: Int): List[Tree] = 1 to n map tupleSel(binder) toList
       private def genTake(binder: Symbol, n: Int): List[Tree]         = (0 until n).toList map (codegen index seqTree(binder))
       private def genDrop(binder: Symbol, n: Int): List[Tree]         = codegen.drop(seqTree(binder))(expectedLength) :: Nil
 
-      // codegen.drop(seqTree(binder))(nbIndexingIndices)))).toList
-      protected def seqTree(binder: Symbol)                = tupleSel(binder)(firstIndexingBinder + 1)
-      protected def tupleSel(binder: Symbol)(i: Int): Tree = {
+      private def seqTree(binder: Symbol): Tree =
+        if (firstIndexingBinder == 0) ref(binder)
```

Comment:
It's overriden in the only trait instance https://github.com/lampepfl/dotty/pull/2157/files#diff-b72950e41046b80e7386c4936d560cf9L1611